gym-unity/gym_unity/envs/__init__.py

```diff
@@ -368,52 +372,69 @@ def _sanitize_info(self, step_result: BatchedStepResult) -> BatchedStepResult:
         # Note that the corresponding agent_id that gets marked done will be different
         # than the original agent that was done, but this is OK since the gym interface
         # only cares about the ordering.
         for index, agent_id in enumerate(step_result.agent_id):
             if not self._previous_step_result.contains_agent(agent_id):
+                # insert the id in the current id list:
+                original_index = self._gym_id_order.index(-1)
+                self._gym_id_order[original_index] = agent_id
+                # This is a new agent
                 step_result.done[index] = True
-            if agent_id in self._done_agents:
-                step_result.done[index] = True
-        self._done_agents = set()
+                # The index of the agent among not-done agents is
+                original_index = self._gym_id_order.index(agent_id)
+                step_result.reward[index] = self._done_agents_index_to_last_reward[
+                    original_index
+                ]
+        self._done_agents_index_to_last_reward = {}
+
         self._previous_step_result = step_result  # store the new original
 
+        new_id_order = []
+        agent_id_list = list(step_result.agent_id)
+        for agent_id in self._gym_id_order:
+            new_id_order.append(agent_id_list.index(agent_id))
+
         _mask: Optional[List[np.array]] = None
         if step_result.action_mask is not None:
             _mask = []
             for mask_index in range(len(step_result.action_mask)):
-                _mask.append(step_result.action_mask[mask_index][indices_to_keep])
+                _mask.append(step_result.action_mask[mask_index][new_id_order])
         new_obs: List[np.array] = []
         for obs_index in range(len(step_result.obs)):
-            new_obs.append(step_result.obs[obs_index][indices_to_keep])
+            new_obs.append(step_result.obs[obs_index][new_id_order])
         return BatchedStepResult(
             obs=new_obs,
-            reward=step_result.reward[indices_to_keep],
-            done=step_result.done[indices_to_keep],
-            max_step=step_result.max_step[indices_to_keep],
-            agent_id=step_result.agent_id[indices_to_keep],
+            reward=step_result.reward[new_id_order],
+            done=step_result.done[new_id_order],
+            max_step=step_result.max_step[new_id_order],
+            agent_id=step_result.agent_id[new_id_order],
             action_mask=_mask,
         )
 
     def _sanitize_action(self, action: np.array) -> np.array:
-        if self._previous_step_result.n_agents() == self._n_agents:
-            return action
         sanitized_action = np.zeros(
             (self._previous_step_result.n_agents(), self.group_spec.action_size)
         )
-        input_index = 0
-        for index in range(self._previous_step_result.n_agents()):
+        for index, agent_id in enumerate(self._previous_step_result.agent_id):
```

Comment:
You don't have to maintain the reverse lookup; you can build before the loop here:
``` python
agent_id_to_gym_index = {agent_id: gym_index for gym_index, agent_id in enumerate(self._gym_id_order)}
for index, agent_id in enumerate(self._previous_step_result.agent_id):
    if not self._previous_step_result.done[index]:
        array_index = agent_id_to_gym_index.get(agent_id, -1)
```